ui/common/src/dialog.ts

```diff
@@ -0,0 +1,223 @@
+          },
+          o.vnodes,
+        ),
+      ),
+    ],
+  );
+}
+
+class DialogWrapper implements Dialog {
+  restoreFocus?: HTMLElement;
+
+  constructor(
+    readonly dialog: HTMLDialogElement,
+    readonly view: HTMLElement,
+    readonly o: DialogOpts,
+  ) {
+    if (dialogPolyfill) dialogPolyfill.registerDialog(dialog); // ios < 15.4
+
+    view.parentElement?.style.setProperty('--vh', `${window.innerHeight * 0.01}px`); // sigh
+    view.addEventListener('click', e => e.stopPropagation());
+
+    dialog.addEventListener('cancel', () => !this.returnValue && (this.returnValue = 'cancel'));
+    dialog.addEventListener('close', this.onClose);
+    dialog
+      .querySelector('.close-button-anchor > .close-button')
+      ?.addEventListener('click', () => this.close('cancel'));
+
+    if (!o.noClickAway) setTimeout(() => document.addEventListener('click', () => this.close('cancel')), 0);
+  }
+
+  get open() {
+    return this.dialog.open;
+  }
+
+  get returnValue() {
+    return this.dialog.returnValue;
+  }
+
+  set returnValue(v: string) {
+    this.dialog.returnValue = v;
+  }
+
+  show = () => {
+    this.returnValue = '';
+    this.dialog.show();
+  };
+
+  close = (v?: string) => {
+    this.dialog.close(this.returnValue || v || 'ok');
+  };
+
+  onClose = () => {
+    if (!this.dialog.returnValue) this.dialog.returnValue = 'cancel';
+    this.o.onClose?.(this);
+    if ('show' in this.o && this.o.show === 'modal') {
+      this.dialog.remove();
+      this.restoreFocus?.focus();
+    }
+    this.restoreFocus = undefined;
+  };
+
+  showModal = () => {
+    this.restoreFocus = document.activeElement as HTMLElement;
+    $(focusQuery, this.view)[1]?.focus();
+    this.view.scrollTop = 0;
+
+    this.addModalListeners?.();
+    this.returnValue = '';
+    this.dialog.showModal();
+  };
+
+  addModalListeners? = () => {
+    this.dialog.addEventListener('keydown', onModalKeydown);
+    this.addModalListeners = undefined; // only do this once per HTMLDialogElement
+  };
+}
+
+function assets(o: DialogOpts) {
+  return Promise.all([
+    o.htmlUrl
+      ? xhr.text(o.htmlUrl)
+      : Promise.resolve(
+          o.cash ? $as<HTMLElement>($(o.cash).clone().removeClass('none')).outerHTML : o.htmlText,
+        ),
+    o.cssPath ? lichess.loadCssPath(o.cssPath) : Promise.resolve(),
+  ]);
+}
+
+function onModalKeydown(e: KeyboardEvent) {
+  if (e.key === 'Tab') {
+    const $focii = $(focusQuery, e.currentTarget as Element),
+      first = $as<HTMLElement>($focii.first()),
+      last = $as<HTMLElement>($focii.last()),
+      focus = document.activeElement as HTMLElement;
+    if (focus === last && !e.shiftKey) first.focus();
+    else if (focus === first && e.shiftKey) last.focus();
+    else return;
+    e.preventDefault();
+  }
+  e.stopPropagation();
+}
+
+function onResize() {
+  const vh = window.innerHeight * 0.01; // ios safari vh behavior not helpful to us
+  $('dialog > div.scrollable').css('--vh', `${vh}px`);
```

Comment:
`vh` sounds like `view height` so I expect it to be just that, in pixels, but here it's divided by 100, why? 
then in the CSS multiplied by 100 again:
```scss
max-height: calc(100 * var(--vh) - 16px);
```

Reply:
`--vh` is actually meant to mimic the css unit `vh` which is given in percents. It's needed because iOS safari has never really agreed with other browsers on what css vh should mean with respect to browser chrome. Multiplying by the --vh variable is explicitly like vh but always with respect to window.innerHeight (similar to all other browsers). But the multiplication / division is not needed. We could just call the variable --availHeight or something